lib/vrl/compiler/src/expression/op.rs

```diff
@@ -143,38 +142,43 @@ impl Expression for Op {
                 (lhs_def - K::Null).merge(rhs_def)
             }
 
-            Or => merged_def,
+            Or => lhs_def.merge(rhs_def),
 
             // ... | ...
-            Merge => merged_def,
+            Merge => lhs_def.merge(rhs_def),
 
             // null && ...
-            And if lhs_kind.is_null() => rhs_def.scalar(K::Boolean),
+            And if lhs_kind.is_null() => rhs_def
+                .fallible_unless(K::Null | K::Boolean)
+                .scalar(K::Boolean),
 
             // ... && ...
-            And => merged_def
+            And => lhs_def
                 .fallible_unless(K::Null | K::Boolean)
+                .merge(rhs_def.fallible_unless(K::Null | K::Boolean))
                 .scalar(K::Boolean),
 
             // ... == ...
             // ... != ...
-            Eq | Ne => merged_def.boolean(),
+            Eq | Ne => lhs_def.merge(rhs_def).boolean(),
 
             // ... >  ...
             // ... >= ...
             // ... <  ...
             // ... <= ...
-            Gt | Ge | Lt | Le => merged_def
+            Gt | Ge | Lt | Le => lhs_def
                 .fallible_unless(K::Integer | K::Float)
+                .merge(rhs_def.fallible_unless(K::Integer | K::Float))
                 .scalar(K::Boolean),
 
             // ... / ...
-            Div => merged_def.fallible().float(),
+            Div => lhs_def.merge(rhs_def).fallible().float(),
```

Comment:
Does this one not need a `fallible_unless` clause? I would expect it to only work on `Integer`/`Float`, but maybe I'm missing something.

Reply:
This one is _always_ fallible, because of potentially division-by-zero at runtime, so it doesn't matter what the type definition is.